src/graphics/device-resource-cache.js

```diff
@@ -0,0 +1,87 @@
+/** @typedef {import('../scene/materials/material.js').Material} Material */
+/** @typedef {import('./texture.js').Texture} Texture */
+/** @typedef {import('./vertex-buffer.js').VertexBuffer} VertexBuffer */
+/** @typedef {import('./graphics-device.js').GraphicsDevice} GraphicsDevice */
+
+/**
+ * A cache storing shared resources associated with a device. The resoruces are removed
+ * from the cache when the device is destroyed.
+ *
+ * @ignore
+ */
+class DeviceResourceCache {
+    /**
+     * An instance of a material which is used when a material is missing.
+     *
+     * @type {Material}
+     */
+    defaultMaterial;
+
+    /**
+     * An instance of a default parameter texture used by the particle emitter.
+     *
+     * @type {Texture}
+     */
+    particleEmitterDefaultParamTexture;
+
+    /**
+     * Vertex buffer containing a quad mesh used by post-effects.
+     *
+     * @type {VertexBuffer}
+     */
+    postEffectQuadVB;
+
+    /** Textures created by reproject-texture module */
+    reprojectTextureCache;
+
+    /** @type {Map<GraphicsDevice, DeviceResourceCache} */
+    static _cache = new Map();
+
+    /** Delete resources, called when the device is destroyed */
+    destroy() {
+        if (this.defaultMaterial) {
+            this.defaultMaterial.destroy();
+            this.defaultMaterial = null;
+        }
+        if (this.particleEmitterDefaultParamTexture) {
+            this.particleEmitterDefaultParamTexture.destroy();
+            this.particleEmitterDefaultParamTexture = null;
+        }
+
+        if (this.postEffectQuadVB) {
+            this.postEffectQuadVB.destroy();
+            this.postEffectQuadVB = null;
+        }
+
+        if (this.reprojectTextureCache) {
+            this.reprojectTextureCache.destroy();
+            this.reprojectTextureCache = null;
+        }
```

Comment:
Not a huge change, just saves some potential lines using `?`:
```suggestion
            this.defaultMaterial?.destroy();
            this.defaultMaterial = null;
            
            this.particleEmitterDefaultParamTexture?.destroy();
            this.particleEmitterDefaultParamTexture = null;
            
            this.postEffectQuadVB?.destroy();
            this.postEffectQuadVB = null;
            
            this.reprojectTextureCache?.destroy();
            this.reprojectTextureCache = null;
```